sample-app/tests/browser/pageoperator.js

```diff
@@ -0,0 +1,58 @@
+const { BrowserPageWidths } = require("./constants");
+
+/**
+ * Responsible for browser navigation based on given test locations.
+ * Each test location is an object which may contain:
+ * - name: name of this test
+ * - path: additional path to append to base url
+ * - viewport: viewport of the page (if omit, desktop view is used by default)
+ * - commands: a list of actions to perform on the page by the page navigator
+ */
+class PageOperator {
+  /**
+   * @param {PageNavigator} pageNavigator page navigator
+   * @param {import('puppeteer').Page} page A Pupeteer Page
+   * @param {Object[]} testLocations a list of test locations to navigate to
+   */
+  constructor(pageNavigator, page, testLocations) {
+    this._pageNavigator = pageNavigator;
+    this._page = page;
+    this._testLocations = testLocations;
+    this._testLocationIndex = -1;
+  }
+
+  hasNextTestLocation() {
+    return this._testLocationIndex < this._testLocations.length - 1;
+  }
+
+  async nextTestLocation() {
```

Comment:
I'm not crazy about the name `TestLocation`. I don't think it accurately describes the object. The name suggests it's essentially just the URL path for the test. I think we should come up with a more descriptive name. As a casual reader of the code, I would imagine this method just moves me to another page. But, it does much more than that.

Reply:
updated to `testInstruction` instead of `testLocation`